### PR TITLE
FIX: If a sub-device is staged, the entire device is staged.

### PR DESCRIPTION
**Pull request by danielballan**
Consider a device `d` with a subdevice `d.s`.

Before #272, `RE(Count([d, d.s]))` errored because the RunEngine tried to redundantly stage `d.s`. Now that `d.stage()` returns `[d, d.s]`, the RunEngine knows to skip staging `d.s`. But `RE(Count([d.s, d]))` (note the reverse order) still errors. The RunEngine has no way of telling `d` not to stage its child `d.s`. (And that is probably a good thing -- it's not the RunEngine's business.)

Possible solutions:
1. Stop being strict that a device can't be staged redundantly. Downside: the risk of failing to clean up and, say, writing into an old file by mistake.
2. Accept arguments to `stage()` that would affect which child devices get staged.
3. Insist that staging a child causes its parent to staged. Thus, `d.s.stage()` would stage `d` as well and return `[d, d.s]`. Thus the RunEngine knows to skip staging `d`.

This PR is Solution 3. That is to say, you can't stage *part* of a device. That makes sense to me: the device was grouped for a reason. And one could imagine people getting into trouble trying to stage part of an area detector (the file-writer but not the trigger, for example).

Implementation note: The ~~`stage_rachet`~~ [renamed] `_defer_to_parent` flag is way of avoiding this infinite loop: child defers to parent, which tries to stage child, which defers to parent, which tries to stage child.... Credit to @tacaswell for thinking of this pattern.

cc @ambarb @cmazzoli Once this gets merged, you can put detectors into `gs.DETS` in any order.

### CODE REVIEW

File: ophyd/device.py [hunk 1]
```diff
@@ -360,11 +360,19 @@ def stage(self):
...
+                                "to cleanup from last use before staging "
+                                "again.")
         elif self._staged == Staged.partially:
             raise RuntimeError("Device has been partially staged. Maybe the "
                                "most recent unstaging encountered an error "
                                "before finishing. Try unstaging again.")
+            if self.parent is not None:
+                # Defer to self.parent, which then stages self and siblings.
+                try:
+                    return self.parent.stage()
```
Comment (tacaswell) [inline ophyd/device.py]: Won't this cause an infinite loop still?

**Comment by danielballan** [comment]
@tacaswell This gets tricky because child calls parent calls sibling calls parent (again) which is at this point in a "partially staged" stage and raises. Latest force-pushed commit might finally have this right.

**Comment by tacaswell** [comment]
I think the parent needs to set `defer_to_parent` to `False` in each of it's children before it calls them.

**Comment by danielballan** [comment]
When you're right, you're right.

Sent from Outlook Mobile<https://aka.ms/qtex0l>




On Sat, Feb 27, 2016 at 5:24 PM -0800, "Thomas A Caswell" <notifications@github.com<mailto:notifications@github.com>> wrote:


I think the parent needs to set defer_to_parent to False in each of it's children before it calls them.

**Comment by klauer** [comment]
The PR message makes me think I understand this, but then I look at the code and end up confused. @danielballan can we have a quick chat about this tomorrow?

**Comment by danielballan** [comment]
@klauer Do the new tests help? I also corrected the implementation. I'm happy to discuss in person tomorrow as well.

File: ophyd/device.py [hunk 2]
```diff
@@ -422,7 +437,12 @@ def unstage(self):
         for attr in self._sub_devices[::-1]:
             device = getattr(self, attr)
             if hasattr(device, 'unstage'):
-                device.unstage()
+                try:
+                    device._defer_unstage_to_parent = False
+                    device.unstage()
+                except Exception:
```
Comment (tacaswell) [inline ophyd/device.py]: I think this should be a `finally`

**Comment by danielballan** [comment]
More tests! In an ideal world, these tests would have less repeated code. I know how to do this in nose, but ten minutes of reading didn't make it clear how to do this in py.test. I'm hoping we can punt and just be grateful for copious tests. :- D

**Comment by tacaswell** [comment]
@ericdill any advice on how to do generative tests?

**Comment by ericdill** [comment]
@tacaswell I assume that by "generative tests" you are talking about the `yield func, arg1, arg2, ...` syntax in nose. Here is a basic example that uses the `@pytest.mark.parameterize` fixture.  @danielballan If I can see your implementation in nose I can show you how I would translate it to pytest

```python
################# Yield style with nose #################

def test_func(num, char):
    # do something interesting with num and char
    pass

def yield_test():
    list1 = [1, 2, 3]
    list2 = list('abc')

    for num, char in zip(list1, list2):
        yield test_func, num, char


################# Fixture style with pytest #################

@pytest.mark.parametrize('num char', zip([1, 2, 3], ['a', 'b', 'c']))
def test_func(num, char):
    # do something interesting with num and char
    pass
```

**Comment by klauer** [comment]
(In case anyone else is thinking to click 'merge' here, please wait - we discussed an alternative offline)

**Pull request closed without merging**